### ISSUE TITLE
Astaire failed to trigger resync on cluster resize

### ISSUE THREAD

**Issue by rkd-msw** [Metaswitch/astaire#78]
#### Symptoms

This was hit on a CCv9 system, and I haven't seen it on CCv10. It may just be a rare race condition, but if there's a change that's likely to have fixed this, I'm happy to close.

After adding a second Sprout node to a cluster, it stayed in a mid-scale-up state overnight:

```
clearwater@sprout-1:~$ cat /etc/clearwater/cluster_settings
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
#
# WARNING - THIS FILE IS GENERATED BY ETCD AND SHOULD NOT BE EDITED DIRECTLY
#
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
servers=173.16.0.5:11211
new_servers=173.16.0.13:11211,173.16.0.5:11211
clearwater@sprout-1:~$ date
Thu Jan 12 13:26:08 GMT 2017
clearwater@sprout-1:~$ ls -l /etc/clearwater/cluster_settings
-rw-r--r-- 1 root root 315 Jan 11 22:31 /etc/clearwater/cluster_settings
```

It looks like this is because `service astaire wait-sync` is hanging:

```
clearwater@sprout-1:~$ ps aux | grep wait-sync
clearwa+  3652  0.0  0.0  10468  2188 pts/4    S+   13:28   0:00 grep --color=auto wait-sync
root      7815  0.0  0.0   4448   856 ?        S    Jan11   0:00 /bin/sh -c service astaire wait-sync
root      7818  0.0  0.0   4448  1768 ?        S    Jan11   0:05 /bin/sh /etc/init.d/astaire wait-sync
```

However, this isn't a duplicate of #76 as far as I can tell - Astaire is running, and was running when clearwater-cluster-manager ran wait-sync:

```
clearwater@sprout-1:~$ ps aux | grep /usr/share/clearwater/bin/astaire
clearwa+  4697  0.0  0.0  10468  2184 pts/4    S+   13:29   0:00 grep --color=auto /usr/share/clearwater/bin/astaire
astaire   6706  0.0  0.1 882496  8064 ?        SNl  Jan11   0:07 /usr/share/clearwater/bin/astaire --local-name=173.16.0.13:11211 --cluster-settings-file=/etc/clearwater/cluster_settings --log-file=/var/log/astaire --log-level=2 --pidfile=/var/run/astaire/astaire.pid
```

/var/log/clearwater-cluster-manager/cluster-manager_20170111T220000Z.txt has this log:

`11-01-2017 22:31:36.978 UTC INFO synchronization_fsm.py:49 (thread MemcachedPlugin): Calling plugin method MemcachedPlugin.on_new_cluster_config_ready`

and `/var/log/astaire/astaire_20170111T220000Z.txt` says Astaire was running at 22:31:36.978, although only just:

```
11-01-2017 22:31:32.841 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
11-01-2017 22:31:32.841 UTC Status memcached_config.cpp:109:  servers=$[signaling_ip_qual]:11211
11-01-2017 22:31:35.010 UTC Status main.cpp:230: Log level set to 2
11-01-2017 22:31:35.010 UTC Status main.cpp:258: Astaire starting up
11-01-2017 22:31:35.021 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
11-01-2017 22:31:35.022 UTC Status memcached_config.cpp:109:  servers=$[signaling_ip_qual]:11211
11-01-2017 22:31:37.008 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
11-01-2017 22:31:37.008 UTC Status memcached_config.cpp:109:  servers=173.16.0.5:11211
11-01-2017 22:31:37.008 UTC Status memcached_config.cpp:109:  new_servers=173.16.0.13:11211,173.16.0.5:11211
11-01-2017 22:31:37.009 UTC Status alarm.cpp:72: astaire issued 5501.5 alarm
11-01-2017 22:32:05.011 UTC Status alarm.cpp:72: astaire issued 5501.5 alarm
11-01-2017 22:32:35.011 UTC Status alarm.cpp:72: astaire issued 5501.5 alarm
```

There are still buckets to resync, so wait-sync is correctly waiting:

```
clearwater@sprout-1:~$ /usr/share/clearwater/astaire/bin/cw_stat astaire astaire_global
bucketsNeedingResync:256
bucketsResynchronized:0
entriesResynchronized:0
dataResynchronized:0
bandwidth:0
```

#### Impact

Scale-up never completes (or at least needs manual intervention to do so).

#### Release and environment 

9.0.0-2-1.0.0-160629.114408

#### Steps to reproduce

Unknown.

**Comment by rkd-msw** [comment]
Diags at \\ad\csd\bigfiles\clearwater\issue-attachments\astaire\78

**Comment by cswindle** [comment]
I have hit this again in a deployment that I created last night:

10.225.118.65

We have sent this image to a customer to prepare for a demo of scale out at MWC, so this is a bit concerning that we have hit this twice, especially as we have not performed very many tests on scale out. Could you collect any logs and perform any diagnosis you require on the live system so that we can continue our automated testing.

**Comment by rkd-msw** [comment]
@cswindle , just to record what we discussed - https://github.com/Metaswitch/astaire/pull/77 isn't a sufficient fix here, because there is actually subscriber data on this system that needs resyncing.

**Comment by mirw** [comment]
Just for the record, both of these were on old releases - we haven't (yet) seen this on the latest code.

**Comment by mirw** [comment]
Astaire logs show:

```
16-01-2017 22:05:25.931 UTC Status main.cpp:230: Log level set to 2
16-01-2017 22:05:25.931 UTC Status main.cpp:258: Astaire starting up
16-01-2017 22:05:25.932 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
16-01-2017 22:05:25.932 UTC Status memcached_config.cpp:109:  servers=$[signaling_ip_qual]:11211
16-01-2017 22:05:42.626 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
16-01-2017 22:05:42.626 UTC Status memcached_config.cpp:109:  servers=173.16.0.9:11211
16-01-2017 22:05:42.626 UTC Status memcached_config.cpp:109:  new_servers=173.16.0.13:11211,173.16.0.9:11211
16-01-2017 22:05:42.627 UTC Status alarm.cpp:72: astaire issued 5501.5 alarm
16-01-2017 22:05:55.931 UTC Status alarm.cpp:72: astaire issued 5501.5 alarm
16-01-2017 22:06:25.931 UTC Status alarm.cpp:72: astaire issued 5501.5 alarm
```

Not sure what that `signaling_ip_qual` bit is about - looks like a misconfiguration, but it should be cleared when we reload configuration.  The alarm (https://github.com/Metaswitch/astaire/blob/dev/root/usr/share/clearwater/infrastructure/alarms/astaire_alarms.json#L40) indicates:

```
                    "severity": "MINOR",
                    "details": "Astaire is resynchronizing the local Memcached process with the rest of the cluster. Service should be unaffected. If a scale-in or scale-out operation is in progress, wait for resynchronization to finish before completing the operation.",
                    "description": "Astaire: Memcached database resynchronization in progress.",
                    "cause": "Astaire is resynchronizing the local Memcached process with the rest of the cluster.",
                    "effect": "Service is unaffected.",
                    "action": "If a scale-in or scale-out operation is in progress, wait for resynchronization to finish before completing the operation."
```

Statistics show that we have 256 buckets needing resynchronization but no connections:

```
$ /usr/share/clearwater/astaire/bin/cw_stat astaire astaire_global
bucketsNeedingResync:256
bucketsResynchronized:0
entriesResynchronized:0
dataResynchronized:0
bandwidth:0
$ /usr/share/clearwater/astaire/bin/cw_stat astaire astaire_connections
$
```

So, did we never try to resynchronize, or has something corrupted our count of buckets needing resync?

**Comment by mirw** [comment]
I deleted `astaire\\tag` from memcached using `delete astaire\\tag`.  This is the tag by which astaire detects that memcached hasn't restarted, and deletion should trigger a resync - it didn't.  This suggests that astaire's control_thread is stuck in the `do_resync` at https://github.com/Metaswitch/astaire/blob/dev/src/astaire.cpp#L188 - nothing else should block, and this means that we should go round and reset the flag after 10s.

This is confirmed by ENT logs - we see a CL_ASTAIRE_START_RESYNC log without a corresponding CL_ASTAIRE_COMPLETE_RESYNC log.

OK, so we're stuck in `process_worklist`.

**Comment by mirw** [comment]
Stack dump shows:

```
Thread 5 (Thread 0x7f8040ff9700 (LWP 5563)):
#0  0x00007f804c96a66b in pthread_join () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000000000041833f in Astaire::complete_single_tap (this=this@entry=0x1d44560, thread_id=<optimized out>, tap_server="") at astaire.cpp:698
Python Exception <class 'IndexError'> list index out of range:
#2  0x000000000041b286 in Astaire::process_worklist (this=this@entry=0x1d44560, owl=std::map with 128 elements) at astaire.cpp:600
#3  0x000000000041b9c9 in Astaire::do_resync (this=this@entry=0x1d44560, full_resync=<optimized out>) at astaire.cpp:480
#4  0x000000000041babb in Astaire::control_thread (this=0x1d44560) at astaire.cpp:188
#5  0x000000000041bbd9 in Astaire::control_thread_fn (data=<optimized out>) at astaire.cpp:218
#6  0x00007f804c969182 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#7  0x00007f804b90247d in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

...and it looks like we're waiting on:

```
Thread 2 (Thread 0x7f802affd700 (LWP 10319)):
#0  0x00007f804c9707eb in recv () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x000000000040856a in recv (__flags=0, __n=16384, __buf=0x7f802aff8cd0, __fd=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/socket2.h:44
#2  Memcached::Connection::recv (this=this@entry=0x7f802affcdc0, msg=msg@entry=0x7f802affcd78) at memcached_tap_client.cpp:411
#3  0x00000000004193c2 in Astaire::tap_buckets_thread (data=0x7f8024002200) at astaire.cpp:267
#4  0x00007f804c969182 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#5  0x00007f804b90247d in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

Interestingly, `tap_server` = "".  Maybe we've failed to connect immediately and that's why we've got stuck?

**Comment by mirw** [comment]
The `tap_server` bit is a red herring - this is an output parameter.

**Comment by mirw** [comment]
The debugger shows that Memcached::Connection::recv is receiving on file descriptor 25.

```
#2  Memcached::Connection::recv (this=this@entry=0x7f802affcdc0, msg=msg@entry=0x7f802affcd78) at memcached_tap_client.cpp:411
411     memcached_tap_client.cpp: No such file or directory.
(gdb) print *this
$1 = {_address = "173.16.0.9:11211", _sock = 25, _buffer = ""}
```

`lsof` shows that this is established:

```
astaire 5549 astaire   25u  IPv4              37974      0t0    TCP 173.16.0.13:52326->173.16.0.9:11211 (ESTABLISHED)
```

So why are we not receiving on it?

**Comment by mirw** [comment]
I'm not sure there's too much more we can do here.  I think the short-term fix is to impose a timeout on this receive, and to improve diagnostics when we receive messages (in particular logging messages we don't understand, which we don't do at present).

**Comment by mirw** [comment]
@eleanor-merry, please can you triage when you're back?

**Comment by cswindle** [comment]
With the timeout code it looks like it now recovers:

18-01-2017 06:09:15.390 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
18-01-2017 06:09:15.390 UTC Status memcached_config.cpp:109:  servers=173.16.0.7:11211
18-01-2017 06:09:15.390 UTC Status memcached_config.cpp:109:  new_servers=173.16.0.13:11211,173.16.0.7:11211
18-01-2017 06:09:15.390 UTC Status memcached_backend.cpp:137: Updating memcached store configuration
18-01-2017 06:09:15.391 UTC Status memcached_backend.cpp:158: Finished preparing new view, so flag that workers should switch to it
18-01-2017 06:09:15.391 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
18-01-2017 06:09:15.391 UTC Status memcached_config.cpp:109:  servers=173.16.0.7:11211
18-01-2017 06:09:15.391 UTC Status memcached_config.cpp:109:  new_servers=173.16.0.13:11211,173.16.0.7:11211
18-01-2017 06:09:15.392 UTC Status alarm.cpp:62: astaire issued 5501.5 alarm
18-01-2017 06:09:25.392 UTC Error memcached_tap_client.cpp:475: Error during recv() on socket (11: Resource temporarily unavailable)
18-01-2017 06:09:25.392 UTC Error astaire.cpp:644: Failed to stream some buckets
18-01-2017 06:09:25.392 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:09:40.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:10:10.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:10:40.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:11:10.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:11:40.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:12:10.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:12:40.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:13:10.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:13:40.637 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:14:10.638 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm
18-01-2017 06:14:35.148 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
18-01-2017 06:14:35.148 UTC Status memcached_config.cpp:109:  servers=173.16.0.13:11211,173.16.0.7:11211
18-01-2017 06:14:35.148 UTC Status memcached_config.cpp:109:  new_servers=173.16.0.13:11211,173.16.0.14:11211,173.16.0.7:11211
18-01-2017 06:14:35.149 UTC Status memcached_config.cpp:70: Reloading memcached configuration from '/etc/clearwater/cluster_settings'
18-01-2017 06:14:35.149 UTC Status memcached_config.cpp:109:  servers=173.16.0.13:11211,173.16.0.7:11211
18-01-2017 06:14:35.149 UTC Status memcached_config.cpp:109:  new_servers=173.16.0.13:11211,173.16.0.14:11211,173.16.0.7:11211
18-01-2017 06:14:35.149 UTC Status memcached_backend.cpp:137: Updating memcached store configuration
18-01-2017 06:14:35.149 UTC Status memcached_backend.cpp:158: Finished preparing new view, so flag that workers should switch to it
18-01-2017 06:14:40.638 UTC Status alarm.cpp:62: astaire issued 5501.1 alarm

**Comment by bossmc** [comment]
From those logs we can see that we're hitting the timeout (the "Resource Temporarily Unavailable" error) while performing a resync, but it's not clear (in part because someone hasn't applied code review markups) at what point the resync occurred.  There are three places I can think of:

* Waiting for a `TAP_INIT` response
" Waiting for a `TAP_MUTATE` during the stream
* Waiting for the connection `close()` after all keys are streamed

In either of the first two cases, we'd expect to see the new node having incomplete data, and that scaling up multiple times (and hitting this issue each time) would slowly degrade the integrity of the database, leading to call failures.  This isn't happening in our tests, which means one of two things:

* We're hitting the timeout at the end of the stream of `TAP_MUTATE` messages in which case it's benign
* There's something else restoring redundancy (maybe after the error we don't find the `astaire\\tag` entry in the local memcached and thus perform a recovery?) which is shrinking the window for data-loss so we don't see any errors, but the window still exists.

**Comment by eleanor-merry** [comment]
I've merged the PR in now. I'm decriticalising this as the testing suggests we're not losing any data.
